src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndex.cs

```diff
@@ -12,65 +12,62 @@ internal class FunctionIndex : IFunctionIndex, IFunctionIndexCollector
     {
         private readonly IDictionary<string, IFunctionDefinition> _functionsById;
         private readonly IDictionary<MethodInfo, IFunctionDefinition> _functionsByMethod;
+        private readonly IDictionary<string, IFunctionDefinition> _functionsByName;
         private readonly ICollection<FunctionDescriptor> _functionDescriptors;
 
         public FunctionIndex()
         {
             _functionsById = new Dictionary<string, IFunctionDefinition>();
             _functionsByMethod = new Dictionary<MethodInfo, IFunctionDefinition>();
+            _functionsByName = new Dictionary<string, IFunctionDefinition>();
             _functionDescriptors = new List<FunctionDescriptor>();
         }
 
         public void Add(IFunctionDefinition function, FunctionDescriptor descriptor, MethodInfo method)
         {
-            string id = descriptor.Id;
-
-            if (_functionsById.ContainsKey(id))
+            if (_functionsById.ContainsKey(descriptor.Id))
             {
-                throw new InvalidOperationException("Method overloads are not supported. " +
-                    "There are multiple methods with the name '" + id + "'.");
+                throw new InvalidOperationException($"Method overloads are not supported. There are multiple methods with the name '{descriptor.Id}'.");
             }
 
-            _functionsById.Add(id, function);
+            _functionsById.Add(descriptor.Id, function);
             _functionsByMethod.Add(method, function);
+
+            // For compat, accept either the short name ("Class.Name") or log name (just "Name")
+            _functionsByName.Add(descriptor.LogName, function);
+            if (descriptor.ShortName != descriptor.LogName)
+            {
+                _functionsByName.Add(descriptor.ShortName, function);
+            }
+
             _functionDescriptors.Add(descriptor);
         }
 
         public IFunctionDefinition Lookup(string functionId)
         {
-            if (!_functionsById.ContainsKey(functionId))
+            if (!_functionsById.TryGetValue(functionId, out IFunctionDefinition function))
             {
                 return null;
             }
-
-            return _functionsById[functionId];
+            return function;
         }
 
         public IFunctionDefinition LookupByName(string name)
         {
-            // For compat, accept either the short name ("Class.Name") or log name (just "Name")
-            foreach (var items in _functionDescriptors)
```

Comment:
This is the overload used by Functions Host, and it resulted in a linear lookup each time.